pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstance.java

```diff
@@ -158,9 +177,25 @@ private void processAsyncResults(BiConsumer<Record, JavaExecutionResult> resultC
     }
 
     @Override
-    public void close() {
+    public void close() throws Exception {
         context.close();
         executor.shutdown();
+        if (null != function && function instanceof HookFunction) {
+            try {
+                ((HookFunction) function).cleanup();
+            } catch (Exception e) {
+                log.error("function closeResource occurred exception", e);
+                throw e;
+            }
+        }
+        if (null != javaUtilFunction && javaUtilFunction instanceof HookFunction) {
+            try {
+                ((HookFunction) javaUtilFunction).cleanup();
+            } catch (Exception e) {
```

Comment:
I suggest we keep `java.util.Function` as simple as possible, and not add hook method for it

Reply:
I think HookFunction can provide users with additional choices, and it also makes the responsibilities of each method of Function more single and clear, please think again